ENH: Specialize some np.char functions to operate on raw C data

lysnikolaou opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

lysnikolaou committed:
Move vec_string related stuff to new file

lysnikolaou committed:
Start working on infrastructure to allow specialization of funcs

lysnikolaou committed:
Implement isalpha for byte strings

lysnikolaou committed:
Add unicode version for isalpha

lysnikolaou committed:
Get itemsize in the specialized function

lysnikolaou committed:
Add infra for add (functions with args)

mattip commented:
`np.char.xxx` functions operate on byte- or unicode- ndarrays. They iterate across the items in the array, applying the `xxx` operation on each item. The current code converts each item to a python `bytes` or `str` object, performs the operation, and then writes the contents back to the result array item. It should be faster to avoid the conversion and operate directly on the byte array/usc32 string. The technique here sets up an array of fast functions, and will use the fast alternative if one exists.

@lysnikolaou to prove the performance improves, could you run the appropriate `spin` command to generate a comparison (hopefully some `np.char` benchmarks already exist)?

lysnikolaou committed:
Remove fastop functions and use argument for fastops instead

lysnikolaou commented:
Here's a couple of benchmarks I wrote (unfortunately there aren't any right now).

```python3
class NumPyChar(Benchmark):
    def setup(self):
        self.A = np.array([100*'x', 100*'y'])
        self.B = np.array([100*'z', 100*'p'])

    def time_add(self):
        np.char.add(self.A, self.B)

    def time_isalpha(self):
        np.char.isalpha(self.A)
```

And the results:

```
All benchmarks:

| Change   | Before [70834eb3] <main>   | After [ad0ceebb] <vec-string-optimization>   |   Ratio | Benchmark (Parameter)             |
|----------|----------------------------|----------------------------------------------|---------|-----------------------------------|
| -        | 17.6±0.5μs                 | 9.09±0.01μs                                  |    0.52 | bench_core.NumPyChar.time_add     |
| -        | 7.47±0.09μs                | 2.97±0.01μs                                  |    0.4  | bench_core.NumPyChar.time_isalpha |
```

ngoldbaum commented:
Cool! Have you checked if this works with [stringdtype](https://github.com/numpy/numpy-user-dtypes/tree/main/stringdtype)? There are tests in that repo to make sure all the ufunc-likes in `np.char` works.

I have some vague ideas about making these a special kind of ufunc but if we can get some big improvements without doing that, even better.

mattip commented:
Would it be possible to avoid converting the method output to a `PyObject*` which is then assigned to the output ndarray? After all you have access to the raw data of the output just like you have access to the raw data of the input.

lysnikolaou committed:
Write result to raw data instead of creating a PyObject

lysnikolaou commented:
> Would it be possible to avoid converting the method output to a `PyObject*` which is then assigned to the output ndarray?

Done in https://github.com/numpy/numpy/pull/24756/commits/462191b3bf455588807d2b4c15a2326f3a295409.

mattip commented:
Could you add your benchmarks to benchmarks/benchmarks and run them again?

lysnikolaou commented:
> Could you add your benchmarks to benchmarks/benchmarks and run them again?

Should I push these to this PR as well?

lysnikolaou commented:
And here's the results with bigger lists!

Code:
```python3
class NumPyChar(Benchmark):
    def setup(self):
        self.A = np.array([100*'x', 100*'y'])
        self.B = np.array([100*'z', 100*'p'])
        self.C = np.array(1000 * ['aa'])
        self.D = np.array(1000 * ['bb'])

    def time_add_small_list_big_string(self):
        np.char.add(self.A, self.B)

    def time_isalpha_small_list_big_string(self):
        np.char.isalpha(self.A)

    def time_add_big_list_small_string(self):
        np.char.add(self.C, self.D)

    def time_isalpha_big_list_small_string(self):
        np.char.isalpha(self.C)
```

Results:
```
| Change   | Before [70834eb3] <main>   | After [462191b3] <vec-string-optimization>   |   Ratio | Benchmark (Parameter)                                   |
|----------|----------------------------|----------------------------------------------|---------|---------------------------------------------------------|
| -        | 7.70±0.05μs                | 2.90±0.06μs                                  |    0.38 | bench_core.NumPyChar.time_isalpha_small_list_big_string |
| -        | 18.0±0.2μs                 | 6.67±0.1μs                                   |    0.37 | bench_core.NumPyChar.time_add_small_list_big_string     |
| -        | 1.35±0.01ms                | 13.5±0.09μs                                  |    0.01 | bench_core.NumPyChar.time_isalpha_big_list_small_string |
| -        | 3.39±0.06ms                | 16.2±0.3μs                                   |    0    | bench_core.NumPyChar.time_add_big_list_small_string     |
```

lysnikolaou commented:
> Cool! Have you checked if this works with [stringdtype](https://github.com/numpy/numpy-user-dtypes/tree/main/stringdtype)? There are tests in that repo to make sure all the ufunc-likes in `np.char` work with it.
> 
> I have some vague ideas about making these a special kind of ufunc but if we can get some big improvements without doing that, even better.

This currently segfaults with `stringdtype`. Looking into it.

mattip commented:
If string dtype seems hard, we can do that in a follow-on PR. 

> Should I push these to this PR as well?

Yes please. Those results seem very encouraging. Could you add a release note?

ngoldbaum commented:
Is there a rush to get this merged? I'd appreciate it if we don't intentionally break stringdtype on main. I'm happy to help out with a fix.

lysnikolaou commented:
> Could you add a release note?

Should we maybe add a release note when the other functions have been switched over to this implementation as well? Otherwise, should I just say `add` and `isalpha` have been re-implemented?

mattip commented:
In the interest of limiting the scope of this first PR, to get it into the hands of users:

I think we can have a "fast path" for `np.bytes_` and `np.str_` only for now, and later rethink the `np.char` mechanism. That will not improve the performance for the new string dtype, but will also not break it. As @ngoldbaum says, there may be some overhaul that sees the common functions become full-fledged ufuncs.

As for the release note, I think it should simply state what has been done. If we go further we can edit the note or add a new one. There are over 60 functions in the `np.char` namespace, and not all of them are worthy of conversion.

lysnikolaou committed:
Use slow path for user defined string arrays

lysnikolaou committed:
Add benchmarks for np.char functions

lysnikolaou committed:
Add release note

lysnikolaou commented:
> I think we can have a "fast path" for np.bytes_ and np.str_ only for now

Changed in https://github.com/numpy/numpy/pull/24756/commits/21f3a2ae46d2bfd5e075f569f0ba1a9f62b4b024.

I also added a release note and pushed the benchmarks I used, so this appears to be ready to be merged.

ngoldbaum commented:
Cool, I can take a look at what's needed to bring these improvements to stringdtype once it's merged.

lysnikolaou committed:
Add comment for static array of fast ops + save release file

lysnikolaou committed:
Add get_length helper and use that in all of the functions

lysnikolaou committed:
Add a stringlib that abstracts string ops out of vec_string.c

mattip commented on numpy/core/src/multiarray/vec_string.c:
```diff
@@ -0,0 +1,365 @@
+    fast_method_found = get_fast_op(char_array, method_name, &fast_method, &fast_method_with_args);
+    if (fast_method_found && !user_defined_dtype) {
+        if (fast_method_with_args) {
+            result = _vec_string_with_args(char_array, type, fast_method, args_seq, /*fast*/ 1);
```
Maybe use variables named like `int fast = 1; int slow = 0` and then use them instead of the comment.

mattip commented on numpy/core/src/multiarray/stringlib/README.txt:
```diff
@@ -0,0 +1,2 @@
+stringlib uses the same design idea (and "templating" mechanism) with the stringlib implementation
+in CPython (https://github.com/python/cpython/tree/main/Objects/stringlib)
```
```suggestion
in CPython (https://github.com/python/cpython/tree/main/Objects/stringlib)

The implementations in the header files are included twice: once after including stringlib.h (for bytes) and once after including ucs4lib.h (for ucf-4 unicode strings).
```

lysnikolaou committed:
Address feedback - use fast variable and improve README text in stringlib

seberg commented on numpy/core/src/multiarray/stringlib/add.h:
```diff
@@ -0,0 +1,40 @@
+    right_len = STRINGLIB(get_length)(in_iters[1]);
+    right_len_in_bytes = right_len * sizeof(STRINGLIB_CHAR);
+
+    if (left_len + right_len > PY_SSIZE_T_MAX) {
```
If this overflows, its negative.  So this doesn't guard against anything, I think?!
You could use `size_t`, I suppose.  (In which case, overflows are not even possible for `add`, since even if we bump limit, they would be bumped to `ssize_t` and `size_t` is at least double that range)

lysnikolaou replied:
Oh yeah, that's correct. Fixed in https://github.com/numpy/numpy/pull/24756/commits/cea8f067360993ed9070db130246cac71ae0545e by not doing the addition at all.

seberg commented on numpy/core/src/multiarray/vec_string.c:
```diff
@@ -0,0 +1,368 @@
+// This static array contains all of the np.char functions that support
+// fast ops, according to the length of the method name as a string.
+// For example, __add__ and isalpha are at index seven, cause they're
+// seven characters long.
```
In some places what we use is a switch statement based on the first character (and or length?) and only then match I think.  But not sure.  Feels a bit oversizes on first sight, but I am OK with it.

seberg reviewed (commented):
If you are happy, I am fine with this; I did not do a deep review.  The only comment is that the overflow check seemed just wrong.

I will admit that I don't really like the direction, because it would be nicer if these could move towards being ufuncs (even fully internal ufuncs!) instead.  We already have `equal` work for strings, there is nothing stopping us (yes, it is a bit more boilerplate).

lysnikolaou committed:
Address feedback - fix overflow check

lysnikolaou commented:
I've already got implementations ready for `find` and `rfind`. Should we merge this as-is or should I switch gears and look into making all of these into ufuncs?

ngoldbaum commented:
If there's ongoing effort going into optimizing these and it's doable to make them ufuncs without too much extra pain, I would vote for making them ufuncs. It would be easier for me to support them in `stringdtype` if they were ufuncs. That's not a veto to continuing along this route, just me advocating for something that would make integrating these changes easier for me in the future.

mattip commented:
Why would writing the inner loop for a ufunc be easier than writing the inner loop for the stringlib utility library? These feel like separate functions not amenable to ufunc generalization:
- in most cases the functions do not generalize to integer/float loops
- there is no need for axis handling nor need for a reduction form of the function
- providing an "out" kwarg will be complicated: what happens if the output width does not match? Error? Truncate the result? Teaching the behaviour of `out` could be difficult.

ngoldbaum commented:
I don't think `out` is particularly confusing, you'd replace any existing string data with the result, just like other ufuncs.

I agree these are special string-specific ufuncs, that's one reason I proposed `np.strings` in NEP 55, as a place in the API to put these ufuncs and to emphasize they're peculiar to strings.

It doesn't make much difference to a dtype implemented inside numpy, but for an external dtype it's much more straightforward to get a ufunc working than an internal function. If NEP 55 is accepted it doesn't make much difference to me personally, but if someone else wants to write a different string dtype for whatever reason, they'll have the same issue.

ngoldbaum commented:
Oh, for fixed-width strings I think `out` would truncate, but IMO this is just another usability issue with fixed-width strings that goes away with stringdtype.

mattip commented:
@lysnikolaou we discussed this at the community meeting. The general sentiment was that it would be better to implement the fast options as ufuncs, rather than in a stringlib. It is a bit more work but would generalize to more dtypes and allow others to write inner loops.

lysnikolaou commented:
Alright, I'm gonna get to work on writing the ufuncs for `isalpha` and `add` then. I'll need some extra learning time, cause it'll be my first time writing ufuncs though.

ngoldbaum commented:
I'm going to close this in favor of the ufunc-based approach. Please feel free to reopen if this was closed in error.
